WalletWasabi.Backend/Controllers/BlockchainController.cs

```diff
@@ -140,76 +143,107 @@ private async Task<IEnumerable<string>> GetRawMempoolStringsNoCacheAsync(Cancell
 		}
 
-		var parsedIds = new List<uint256>();
+		uint256[] parsedTxIds;
+
+		// Make sure TXIDs are not malformed.
 		try
 		{
-			// Remove duplicates, do not use Distinct(), order is not guaranteed.
-			foreach (var txid in transactionIds.Select(x => new uint256(x)))
-			{
-				if (!parsedIds.Contains(txid))
-				{
-					parsedIds.Add(txid);
-				}
-			}
+			parsedTxIds = transactionIds.Select(x => new uint256(x)).ToArray();
 		}
 		catch
 		{
 			return BadRequest("Invalid transaction Ids.");
 		}
 
-		try
+		if (requestCount == 1)
+		{
+			var singleTx = parsedTxIds.Single();
+			var tx = await GetTransactionAsync(singleTx, cancellationToken);
+			return Ok(new[] { tx });
+		}
+
+		HashSet<uint256> txIdsRetrieve = [];
+		Task<Transaction>?[] tasks = new Task<Transaction>?[requestCount];
+
+		// Fill transactions we already have in cache. Array can look like [null, tx1, null, tx2, etc.].
+		for (int i = 0; i < requestCount; i++)
 		{
-			var hexes = new Dictionary<uint256, string>();
-			List<uint256> missingTxs = new();
-			lock (TransactionHexCacheLock)
+			uint256 txId = parsedTxIds[i];
+			string cacheKey = GetCacheKeyForTransaction(txId);
+
+			if (Cache.TryGetValue(cacheKey, out TaskCompletionSource<Transaction>? tcs))
 			{
-				foreach (var txid in parsedIds)
-				{
-					if (TransactionHexCache.TryGetValue(txid, out string? hex))
-					{
-						hexes.Add(txid, hex);
-					}
-					else
-					{
-						missingTxs.Add(txid);
-					}
-				}
+				tasks[i] = tcs!.Task;
 			}
-
-			if (missingTxs.Count != 0)
+			else
 			{
-				foreach (var tx in await RpcClient.GetRawTransactionsAsync(missingTxs, cancellationToken))
-				{
-					string hex = tx.ToHex();
-					hexes.Add(tx.GetHash(), hex);
-
-					lock (TransactionHexCacheLock)
-					{
-						if (TransactionHexCache.TryAdd(tx.GetHash(), hex) && TransactionHexCache.Count >= 1000)
-						{
-							TransactionHexCache.Remove(TransactionHexCache.Keys.First());
-						}
-					}
-				}
+				txIdsRetrieve.Add(txId);
 			}
+		}
+
+		// Ask to get missing transactions over RPC.
+		Task<Dictionary<uint256, Transaction>> rpcBatchTask = Task.Run(async () =>
+		{
+			IEnumerable<Transaction> txs = await RpcClient.GetRawTransactionsAsync(txIdsRetrieve, cancellationToken).ConfigureAwait(false);
+			return txs.ToDictionary(x => x.GetHash(), x => x);
+		});
 
-			// Order hexes according to the order of the query.
-			var orderedResult = parsedIds.Where(x => hexes.ContainsKey(x)).Select(x => hexes[x]);
-			return Ok(orderedResult);
+		// Cache should be aware that we are retrieving each and every txid right now.
+		foreach (uint256 txId in txIdsRetrieve) 
+		{
+			uint256 txIdCopy = new(txId); // To capture correct variable in the lambda below.
+
+			_ = await Cache.GetCachedResponseAsync(
+				request: GetCacheKeyForTransaction(txId),
+				action: async (string request, CancellationToken token) =>
+				{
+					Dictionary<uint256, Transaction> dictionary = await rpcBatchTask.ConfigureAwait(false);
```

Comment:
I don't get this, help me. It seems like if we miss 5 txs, we will start 5 tasks that is getting all the missing txs - generating 25 requests in total?